src/Interpreters/TemporaryDataOnDisk.h

```diff
@@ -184,10 +204,15 @@ class TemporaryFileStream : boost::noncopyable
 
     Stat stat;
 
+    /// 0 - means that we haven't requested any read, 1 - read from function TemporaryFileStream::read, 2 - 
+    std::atomic_char read_type{0};
```

Comment:
This complicates logic a bit, I assume simple check `!in_reader` in `getReadStream` is enough. Generally speaking, from the first glance it even wouldn't be a big issue if we allowed calling both methods. But as far it's not necessary for now, we can state that it's not supposed to be so and add simple check.